opensearchpy/helpers/signer.py

```diff
@@ -0,0 +1,89 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+#
+# Modifications Copyright OpenSearch Contributors. See
+# GitHub history for details.
+
+import sys
+
+import requests
+
+OPENSEARCH_SERVICE = "es"
+
+PY3 = sys.version_info[0] == 3
+
+if PY3:
+    from urllib.parse import parse_qs, urlencode, urlparse
+
+
+def fetch_url(prepared_request):  # type: ignore
+    """
+    This is a util method that helps in reconstructing the request url.
+    :param prepared_request: unsigned request
+    :return: reconstructed url
+    """
+    url = urlparse(prepared_request.url)
+    path = url.path or "/"
+
+    # fetch the query string if present in the request
+    querystring = ""
+    if url.query:
+        querystring = "?" + urlencode(
+            parse_qs(url.query, keep_blank_values=True), doseq=True
+        )
+
+    # fetch the host information from headers
+    headers = dict(
+        (key.lower(), value) for key, value in prepared_request.headers.items()
+    )
+    location = headers.get("host") or url.netloc
```

Comment:
Are we passing host to headers? We don’t have to make this method generic. Since prepared request is constructed by us, we can make that assumption and pass it to method if required

Reply:
We are not passing host in headers, it's a soft check. I think we can keep it as it is.